theseus/optimizer/nonlinear/trust_region.py

```diff
@@ -0,0 +1,167 @@
+        if self._trusted_step_idx is None:
+            self._trusted_step_idx = (
+                TrustRegion._squared_norm(delta) <= self._trust_region**2
+            )
+
+        # Storing the indices of delta exactly at the trust boundary.
+        # If the step is at the boundary and the predicted error reduction agrees
+        # with real reduction, we can increase the trust region
+        # in self.complete_step()
+        delta, self._at_trust_boundary_idx = self._compute_delta_impl()
+        if self._at_trust_boundary_idx is None:
+            self._at_trust_boundary_idx = (
+                TrustRegion._squared_norm(delta) <= self._trust_region**2
+            )
+        return delta
+
+    @staticmethod
+    def _squared_norm(tensor: torch.Tensor, keepdim: bool = True) -> torch.Tensor:
+        return (tensor**2).sum(dim=1, keepdim=keepdim)
+
+    # Computes m_k(delta) as defined in Eq. (4.2) of the above reference (p. 68)
+    def _predicted_error(
+        self, previous_error: torch.Tensor, delta: torch.Tensor
+    ) -> torch.Tensor:
+        linearization = self.linear_solver.linearization
+
+        # Note that B = |A^t@ A| so that p^tBp := delta^t A^t A delta = ||Adelta||^2
+        Adelta = linearization.Av(delta)
+        grad = -linearization.Atb.squeeze(2)
+        delta_dot_grad = (delta * grad).sum(dim=1)
+
+        return (
+            previous_error
+            + delta_dot_grad
+            + 0.5 * TrustRegion._squared_norm(Adelta, keepdim=False)
+        )
+
+    def _complete_step(
+        self,
+        delta: torch.Tensor,
+        new_err: torch.Tensor,
+        previous_err: torch.Tensor,
+        accept_threshold: float = 0.0,
+        shrink_threshold: float = 0.25,
+        expand_threshold: float = 0.75,
+        shrink_ratio: float = 0.25,
+        expand_ratio: float = 2.0,
+        **kwargs,
+    ) -> Optional[torch.Tensor]:
+        # "err" tensors passed as input refer to the squared norm of the
+        # error vector, as returned by self._error_metric()
+        good_params = (0.0 < shrink_ratio <= 1.0) and (expand_ratio >= 1.0)
+        good_params &= (shrink_threshold < expand_threshold) and (
+            accept_threshold < shrink_threshold
+        )
+        if not good_params:
+            raise ValueError(
+                "Invalid parameters for TrustRegionMethod. "
+                "Values must satisfy <accept/shrink>_threshold < expand_threshold, "
+                "shrink_ratio in (0, 1], and expand_ratio > 1.0."
+            )
+        pred_err = self._predicted_error(previous_err, delta)
+        rho = ((previous_err - new_err) / (previous_err - pred_err)).view(-1, 1)
+        shrink_idx = rho < shrink_threshold
+        self._trust_region = torch.where(
+            shrink_idx, self._trust_region * shrink_ratio, self._trust_region
+        )
+        expand_idx = (rho > expand_threshold) & self._at_trust_boundary_idx
```

Comment:
I'm following Nocedal (see below). I assumed the rationale for this was to be conservative, since your predicted error might be in agreement with the actual error just because the step was actually smaller than the trust region. 

<img width="533" alt="image" src="https://user-images.githubusercontent.com/4759586/205971608-1e53452d-288f-4eaf-96b2-6f331d909374.png">